src/assert.rs

```diff
@@ -359,25 +360,41 @@ impl Assert {
             if expect_success != output.status.success() {
                 let out = String::from_utf8_lossy(&output.stdout).to_string();
                 let err = String::from_utf8_lossy(&output.stderr).to_string();
-                let err: Error = ErrorKind::StatusMismatch(expect_success, out, err).into();
-                bail!(err.chain_err(|| ErrorKind::AssertionFailed(self.cmd.clone())));
+                Err(AssertionError::new(AssertionKind::StatusMismatch))
+                    .context(if expect_success {
+                        "expected to succeed"
+                    } else {
+                        "expected to fail"
+                    })
+                    .context(KeyValueDisplay::new("stdout", QuotedDisplay::new(out)))
+                    .context(KeyValueDisplay::new("stderr", QuotedDisplay::new(err)))
+                    .with_context(|_| KeyValueDisplay::new("command", format!("{:?}", command)))?;
             }
         }
 
         if self.expect_exit_code.is_some() && self.expect_exit_code != output.status.code() {
             let out = String::from_utf8_lossy(&output.stdout).to_string();
             let err = String::from_utf8_lossy(&output.stderr).to_string();
-            let err: Error =
-                ErrorKind::ExitCodeMismatch(self.expect_exit_code, output.status.code(), out, err)
-                    .into();
-            bail!(err.chain_err(|| ErrorKind::AssertionFailed(self.cmd.clone())));
+            Err(AssertionError::new(AssertionKind::ExitCodeMismatch))
+                .context(KeyValueDisplay::new(
+                    "expected",
+                    self.expect_exit_code.expect("is_some already called"),
+                ))
+                .context(KeyValueDisplay::new(
+                    "got",
+                    output.status.code().unwrap_or_default(),
+                ))
+                .context(KeyValueDisplay::new("stdout", QuotedDisplay::new(out)))
+                .context(KeyValueDisplay::new("stderr", QuotedDisplay::new(err)))
+                .with_context(|_| KeyValueDisplay::new("command", format!("{:?}", command)))?;
```

Comment:
I was hoping to use `DebugDisplay` to adapt a `process::Command` into being usable inside of a `Context<D>` but `process::Command` isn't `Send` (I think that was it).

My plan is to modify `assert_cli` into using extension traits for part of the process.  By doing this, I won't have access anymore to the exact parameters the user passes in.  Instead I'll only have a `process::Command`.  The `Debug` for that is `' '.join('"%s"' % arg for arg in cmd)` (sorry, more succinct in python) which might be good enough.

So I'm stuck with losing the original data and just calling `format!` here.

The other odd thing is process::Command isn't `Clone`.

Reply:
Oh, that's hard trade-off. Good errors are something we shouldn't compromise on.

> The other odd thing is process::Command isn't Clone.

Is this because the Builder and the actually running Command thing are the same?

Reply:
> Is this because the Builder and the actually running Command thing are the same?

From what I understand, `Command` is the builder.  The running command is a `Child` or an `Output`.

Granted, from the error I saw with `Send`, its probably directly populating the OS C-based data structure as you go through the builder.